posthog/temporal/batch_exports/http_batch_export.py

```diff
@@ -0,0 +1,361 @@
+        BatchExportField(expression="nullIf(properties, '')", alias="properties"),
+        BatchExportField(expression="toString(distinct_id)", alias="distinct_id"),
+        BatchExportField(expression="elements_chain", alias="elements_chain"),
+    ]
+
+
+class HeartbeatDetails:
+    """This class allows us to enforce a schema on the Heartbeat details.
+
+    Attributes:
+        last_uploaded_part_timestamp: The timestamp of the last part we managed to upload.
+    """
+
+    last_uploaded_part_timestamp: str
+
+    def __init__(self, last_uploaded_part_timestamp: str):
+        self.last_uploaded_part_timestamp = last_uploaded_part_timestamp
+
+    @classmethod
+    def from_activity_details(cls, details) -> "HeartbeatDetails":
+        last_uploaded_part_timestamp = details[0]
+        return HeartbeatDetails(last_uploaded_part_timestamp)
+
+
+@dataclass
+class HttpInsertInputs:
+    """Inputs for HTTP insert activity."""
+
+    team_id: int
+    url: str
+    token: str
+    data_interval_start: str
+    data_interval_end: str
+    exclude_events: list[str] | None = None
+    include_events: list[str] | None = None
+    batch_export_schema: BatchExportSchema | None = None
+
+
+async def maybe_resume_from_heartbeat(inputs: HttpInsertInputs) -> str:
+    """Returns the `interval_start` to use, either resuming from previous heartbeat data or
+    using the `data_interval_start` from the inputs."""
+    logger = await bind_temporal_worker_logger(team_id=inputs.team_id, destination="HTTP")
+
+    interval_start = inputs.data_interval_start
+    details = activity.info().heartbeat_details
+
+    if not details:
+        # No heartbeat found, so we start from the beginning.
+        return interval_start
+
+    try:
+        interval_start = HeartbeatDetails.from_activity_details(details).last_uploaded_part_timestamp
+    except IndexError:
+        # This is the error we expect when there are no activity details as the sequence will be
+        # empty.
+        logger.debug(
+            "Did not receive details from previous activity Excecution. Export will start from the beginning %s",
+            interval_start,
+        )
+    except Exception:
+        # We still start from the beginning, but we make a point to log unexpected errors. Ideally,
+        # any new exceptions should be added to the previous block after the first time and we will
+        # never land here.
+        logger.warning(
+            "Did not receive details from previous activity Excecution due to an unexpected error. Export will start from the beginning %s",
+            interval_start,
+        )
+
+    return interval_start
+
+
+async def post_json_file_to_url(url, batch_file):
+    batch_file.seek(0)
+    async with aiohttp.ClientSession() as session:
+        headers = {"Content-Type": "application/json"}
+        async with session.post(url, data=batch_file, headers=headers) as response:
+            raise_for_status(response)
+            return response
```

Comment:
The [recommended pattern](https://docs.aiohttp.org/en/stable/client_reference.html) is to use a single `ClientSession` for all requests.